File: packages/sim/src/terminals/generators/firewall.ts
import type { MatchQuestion, RulesQuestion } from "../types";
import { base, fill, param, pool, type GenContext } from "./common";

interface ServiceDef {
  key: string;
  port: number;
  name: string;
  description: string;
}

interface PolicyDef {
  id: string;
  text: string;
  allow: string[];
}

interface PortsPool {
  services: ServiceDef[];
  match: { prompt: string; explanation: string; hint: string };
  rules: {
    prompt: string;
    explanation: string;
    hint: string;
    policies: PolicyDef[];
  };
}

/** Associação porta ↔ serviço. */
export function portServiceMatch(ctx: GenContext): MatchQuestion {
  const p = pool<PortsPool>(ctx, "ports");
  const rng = ctx.rng;
  const count = Math.min(param(ctx, "count", 5), p.services.length);
  const chosen = rng.sample(p.services, count);
  const idx = chosen.map((_, i) => i);
  const rightOrder = rng.shuffle(idx);
  return {
    ...base(ctx, "port_service_match"),
    kind: "match",
    prompt: p.match.prompt,
    left: chosen.map((s) => String(s.port)),
    right: rightOrder.map((i) => chosen[i]!.name),
    answer: idx.map((i) => rightOrder.indexOf(i)),
    explanation: p.match.explanation,
    hint: p.match.hint,
    tags: ["ports"],
  };
}

/**
 * Escrita de regras de firewall a partir de uma política em linguagem natural.
 * Todos os serviços exigidos pela política entram na lista; o resto vem sorteado.
 */
export function firewallRules(ctx: GenContext): RulesQuestion {
  const p = pool<PortsPool>(ctx, "ports");
  const P = p.rules;
  const rng = ctx.rng;
  const policy = rng.pick(P.policies);
  const total = param(ctx, "count", 5);

  const required = p.services.filter((s) => policy.allow.includes(s.key));
  const rest = p.services.filter((s) => !policy.allow.includes(s.key));
  const extra = Math.max(1, Math.min(total - required.length, rest.length));
  const ports = rng.shuffle([...required, ...rng.sample(rest, extra)]);

  const allowed = ports.filter((s) => policy.allow.includes(s.key));
  const denied = ports.filter((s) => !policy.allow.includes(s.key));
  const names = (list: ServiceDef[]): string =>
    list.length === 0 ? "—" : list.map((s) => `${s.name} (${s.port})`).join(", ");
  const vars = { policy: `Este host ${policy.text}`, allowNames: names(allowed), denyNames: names(denied) };

  return {
    ...base(ctx, "firewall_rules"),
    kind: "rules",
    prompt: fill(P.prompt, vars),
    // sem tabela de contexto: a própria grade de regras da UI já lista porta, serviço e descrição
    ports: ports.map((s) => ({ port: s.port, service: s.name, description: s.description })),
    answer: { allowed: allowed.map((s) => s.port) },
    explanation: fill(P.explanation, vars),
    hint: P.hint,
    tags: ["firewall"],
  };
}

File: apps/api/app/terminals/generators/firewall.py
"""Portas e regras de firewall — espelho de packages/sim/src/terminals/generators/firewall.ts."""

from __future__ import annotations

from typing import Any

from .common import GenContext, base, fill, param, pool


def port_service_match(ctx: GenContext) -> dict[str, Any]:
    p = pool(ctx, "ports")
    rng = ctx.rng
    services: list[dict[str, Any]] = p["services"]
    count = min(param(ctx, "count", 5), len(services))
    chosen = rng.sample(services, count)
    idx = list(range(len(chosen)))
    right_order = rng.shuffle(idx)
    return {
        **base(ctx, "port_service_match"),
        "kind": "match",
        "prompt": p["match"]["prompt"],
        "left": [str(s["port"]) for s in chosen],
        "right": [chosen[i]["name"] for i in right_order],
        "answer": [right_order.index(i) for i in idx],
        "explanation": p["match"]["explanation"],
        "hint": p["match"]["hint"],
        "tags": ["ports"],
    }


def _names(items: list[dict[str, Any]]) -> str:
    if not items:
        return "—"
    return ", ".join(f"{s['name']} ({s['port']})" for s in items)


def firewall_rules(ctx: GenContext) -> dict[str, Any]:
    p = pool(ctx, "ports")
    rules = p["rules"]
    rng = ctx.rng
    policy = rng.pick(rules["policies"])
    total = param(ctx, "count", 5)

    allow_keys: list[str] = policy["allow"]
    required = [s for s in p["services"] if s["key"] in allow_keys]
    rest = [s for s in p["services"] if s["key"] not in allow_keys]
    extra = max(1, min(total - len(required), len(rest)))
    ports = rng.shuffle([*required, *rng.sample(rest, extra)])

    allowed = [s for s in ports if s["key"] in allow_keys]
    denied = [s for s in ports if s["key"] not in allow_keys]
    variables = {
        "policy": f"Este host {policy['text']}",
        "allowNames": _names(allowed),
        "denyNames": _names(denied),
    }
    return {
        **base(ctx, "firewall_rules"),
        "kind": "rules",
        "prompt": fill(rules["prompt"], variables),
        # sem tabela de contexto: a grade de regras da UI já lista porta, serviço e descrição
        "ports": [{"port": s["port"], "service": s["name"], "description": s["description"]} for s in ports],
        "answer": {"allowed": [s["port"] for s in allowed]},
        "explanation": fill(rules["explanation"], variables),
        "hint": rules["hint"],
        "tags": ["firewall"],
    }
